tools/ddprobe.cpp

```diff
@@ -69,6 +71,123 @@ syncThreadDesktop() {
+ * @param darknessThreshold A floating-point value representing the threshold above which a pixel's RGB values are considered non-dark. The value ranges from 0.0f to 1.0f, with a default value of 0.1f.
+ * @return bool Returns `true` if the frame is determined to be entirely dark, otherwise returns `false`.
+ */
+bool
+is_valid_frame(const D3D11_MAPPED_SUBRESOURCE &mappedResource, const D3D11_TEXTURE2D_DESC &frameDesc, float darknessThreshold = 0.1f) {
+  const uint8_t *pixels = static_cast<const uint8_t *>(mappedResource.pData);
+  const int bytesPerPixel = 4;  // (8 bits per channel, excluding alpha). Factoring HDR is not needed because it doesn't cause black levels to raise enough to be a concern.
+  const int stride = mappedResource.RowPitch;
+  const int width = frameDesc.Width;
+  const int height = frameDesc.Height;
+
+  // Convert the darkness threshold to an integer value for comparison
+  const int threshold = static_cast<int>(darknessThreshold * 255);
+
+  // Iterate over each pixel in the frame
+  for (int y = 0; y < height; ++y) {
+    for (int x = 0; x < width; ++x) {
+      const uint8_t *pixel = pixels + y * stride + x * bytesPerPixel;
+      // Check if any RGB channel exceeds the darkness threshold
+      if (pixel[0] > threshold || pixel[1] > threshold || pixel[2] > threshold) {
+        // Frame is not dark
+        return true;
+      }
+    }
+  }
+  // Frame is entirely dark
+  return false;
+}
+
+/**
+ * @brief Captures and verifies the contents of up to 10 consecutive frames from a DXGI output duplication.
+ *
+ * This function attempts to acquire and analyze up to 10 frames from a DXGI output duplication object (`dup`).
+ * It checks if each frame is non-empty (not entirely dark) by using the `is_valid_frame` function.
+ * If any non-empty frame is found, the function returns `S_OK`.
+ * If all 10 frames are empty, it returns `S_FALSE`, suggesting potential issues with the capture process.
+ * If any error occurs during the frame acquisition or analysis process, the corresponding `HRESULT` error code is returned.
+ *
+ * @param dup A reference to the DXGI output duplication object (`dxgi::dup_t&`) used to acquire frames.
+ * @param device A ComPtr to the ID3D11Device interface representing the device associated with the Direct3D context.
+ * @return HRESULT Returns `S_OK` if a non-empty frame is captured successfully, `S_FALSE` if all frames are empty, or an error code if any failure occurs during the process.
+ */
+HRESULT
+test_frame_capture(dxgi::dup_t &dup, ComPtr<ID3D11Device> device) {
+  for (int i = 0; i < 10; ++i) {
+    std::cout << "Attempting to acquire frame " << (i + 1) << " of 10..." << std::endl;
+
+    ComPtr<IDXGIResource> frameResource;
+    DXGI_OUTDUPL_FRAME_INFO frameInfo;
+    HRESULT status = dup->AcquireNextFrame(500, &frameInfo, &frameResource);
+    auto release_frame = util::fail_guard([&dup]() {
```

Comment:
I think this should created after the `if (FAILED(status)) {` since we could be releasing a non-acquired frame. I dunno if it's safe or not. If it is, then it's fine.

Reply:
As far as I am aware, fail guards are executed when the scope ends so its totally safe doing that.

Reply:
And that the code is always acquiring a frame, regardless if it succeeds or fails, it always needs to release.